File: _DEMO/agent-server/routes/assistant.js
// POST /ask-assistant — collections agent with tool calling.
//
// SSE event types emitted:
//   data: {"status":"..."}   during tool execution
//   data: {"message":"..."}  final answer chunks
//   data: [DONE]

import express from 'express';
import { getAccessToken, callAiCore } from '../lib/ai-core.js';
import { tools as TOOLS } from '../lib/mcp-client.js'; // [CHG-002]
import { buildMessages, handleToolCalls } from '../lib/agent.js';
import { log, err } from '../lib/logger.js';

export const assistantRouter = express.Router();

function makeSenders(res) {
  return {
    chunk:  (text) => res.write(`data: ${JSON.stringify({ message: text })}\n\n`),
    status: (text) => res.write(`data: ${JSON.stringify({ status: text })}\n\n`),
    done:   ()     => res.write('data: [DONE]\n\n'),
    error:  (msg)  => {
      res.write(`data: ${JSON.stringify({ error: msg })}\n\n`);
      res.write('data: [DONE]\n\n');
    },
  };
}

assistantRouter.post('/ask-assistant', async (req, res) => {
  const { message, history = [], context = {} } = req.body;
  if (!message || typeof message !== 'string' || !message.trim())
    return res.status(400).json({ error: 'message is required' });

  const start = Date.now();
  log('[request]', `message="${message.slice(0, 80)}${message.length > 80 ? '…' : ''}"  history=${history.length} turns  context=${JSON.stringify(context)}`);

  res.setHeader('Content-Type',  'text/event-stream');
  res.setHeader('Cache-Control', 'no-cache');
  res.setHeader('Connection',    'keep-alive');
  res.flushHeaders();

  const send = makeSenders(res);

  try {
    const token    = await getAccessToken();
    const messages = buildMessages(history, message);

    const t1start = Date.now();
    const turn1   = await callAiCore(token, { messages, stream: false, tools: TOOLS, tool_choice: 'auto' });
    const choice  = turn1.choices?.[0];
    const usage   = turn1.usage;
    log('[ai-core]', `turn-1  ${Date.now() - t1start}ms  finish=${choice?.finish_reason}  tokens=${usage?.total_tokens ?? '?'}  (prompt=${usage?.prompt_tokens ?? '?'} + completion=${usage?.completion_tokens ?? '?'})`);

    if (choice?.finish_reason === 'tool_calls') {
      const toolNames = choice.message.tool_calls.map((tc) => tc.function.name).join(', ');
      log('[agent]', `tool_calls: ${toolNames}`);
      await handleToolCalls(token, messages, choice, send, context);
    } else {
      const content = choice?.message?.content ?? '';
      log('[agent]', `direct answer  chars=${content.length}`);
      send.chunk(content);
    }

    log('[done]', `total=${Date.now() - start}ms`);
    send.done();
    res.end();
  } catch (e) {
    err('[error]', e.message);
    send.error('An unexpected error occurred. Please try again.');
    res.end();
  }
});

File: _DEMO/agent-server-py/routes/assistant.py
# [C03-F04, C03-F05, C03-F16] POST /ask-assistant — collections agent with SSE streaming.
#
# SSE event types emitted:
#   data: {"status":"..."}   during tool execution
#   data: {"message":"..."}  final answer chunks
#   data: {"error":"..."}    on unexpected error
#   data: [DONE]

import json
import time

from fastapi import APIRouter, Request
from fastapi.responses import JSONResponse, StreamingResponse

from lib import mcp_client
from lib.agent import build_messages, handle_tool_calls
from lib.ai_core import get_access_token, call_ai_core
from lib.logger import log, err

router = APIRouter()


@router.post("/ask-assistant")
async def ask_assistant(request: Request):
    body = await request.json()

    # [C03-F04] Validate request body.
    message = body.get("message")
    if not message or not isinstance(message, str) or not message.strip():
        return JSONResponse(status_code=400, content={"error": "message is required"})

    history = body.get("history") or []
    context = body.get("context") or {}
    start = time.monotonic()

    truncated = message[:80] + ("…" if len(message) > 80 else "")
    log("[request]", f"message=\"{truncated}\"  history={len(history)} turns  context={json.dumps(context)}")

    # [C03-F05] Collect all SSE events into a list synchronously, then stream them.
    # Using a queue + create_task is unreliable inside async generators; collecting
    # first then yielding avoids task-scheduling races.
    async def event_generator():
        events: list[str] = []

        def emit(data: str) -> None:
            events.append(data)

        try:
            token = await get_access_token()
            messages = build_messages(history, message)

            t1_start = time.monotonic()
            turn1 = await call_ai_core(token, {
                "messages": messages,
                "stream": False,
                "tools": mcp_client.tools,
                "tool_choice": "auto",
            })
            choice = (turn1.get("choices") or [{}])[0]
            usage = turn1.get("usage", {})
            elapsed = int((time.monotonic() - t1_start) * 1000)
            log(
                "[ai-core]",
                f"turn-1  {elapsed}ms  finish={choice.get('finish_reason')}  "
                f"tokens={usage.get('total_tokens', '?')}  "
                f"(prompt={usage.get('prompt_tokens', '?')} + "
                f"completion={usage.get('completion_tokens', '?')})",
            )

            if choice.get("finish_reason") == "tool_calls":
                tool_names = ", ".join(
                    tc["function"]["name"]
                    for tc in choice["message"].get("tool_calls", [])
                )
                log("[agent]", f"tool_calls: {tool_names}")

                # Collect status events and streamed chunks via callbacks.
                streamed_chunks: list[str] = []

                async def send_chunk(text: str) -> None:
                    emit(json.dumps({"message": text}))
                    streamed_chunks.append(text)

                async def send_status(text: str) -> None:
                    emit(json.dumps({"status": text}))

                send = {"chunk": send_chunk, "status": send_status}
                await handle_tool_calls(token, messages, choice, send, context)

            else:
                content = (choice.get("message") or {}).get("content") or ""
                log("[agent]", f"direct answer  chars={len(content)}")
                emit(json.dumps({"message": content}))

            total = int((time.monotonic() - start) * 1000)
            log("[done]", f"total={total}ms")

        except Exception as exc:
            err("[error]", str(exc))
            events.append(json.dumps({"error": "An unexpected error occurred. Please try again."}))

        events.append("[DONE]")

        for data in events:
            yield f"data: {data}\n\n"

    return StreamingResponse(
        event_generator(),
        media_type="text/event-stream",
        headers={
            "Cache-Control": "no-cache",
            "Connection": "keep-alive",
            "X-Accel-Buffering": "no",
        },
    )
